### DISCUSSION TITLE
怎么在全局请求拦截器获取initialState中的数据

### DISCUSSION THREAD

**Discussion by nikojxie** [umijs/umi#11886] [Q&A]
我想在 requestInterceptors 请求头内添加一个参数，值从initialState获取，这里不在组件内无无法使用useModel，请问有什么其他方式获取吗？

**Comment by nikojxie** [comment]
只能使用 valtio 这种吗

**Reply by fz6m** [reply]
值从 `initialState` 获取的话，本质上还是走的 useModel 也就是 hooks ，而你在请求的时候也是在组件里进行，一般是 `useRequest` 或类似的请求 hook 封装，那在这个请求 hooks 封装里把 initialState 调出来传进去就好了，还是都在 hooks 体系内。

如果想不受约束的话，就用 valtio 或者其他可以脱离 hooks 规则的 react 全局状态管理库吧。

**Comment by MeteorsLiu** [comment]
2024年回答，其实可以在getInitialState时候把initialState放到window上，umijs也不会像nextjs那么复杂，所以拦截器肯定是可以获取到window的，避免又引入一个valtio增大体积

 ```
const fetchUserInfo = async () => {
    try {
      const res = await queryCurrentUser();
      ...
      // 用于在拦截器中检查用户变更的，不要删掉
      window._____initState = res;
      return res;
    } catch (error) {
      redirectLogin();
    }
    return undefined;
  };
```

拦截器中使用
`const initState = window._____initState `

测试过有效
